src/compiler/checker.ts

```diff
@@ -6207,14 +6207,20 @@ namespace ts {
             }
 
             function getNarrowedType(originalType: Type, narrowedTypeCandidate: Type) {
-                // Narrow to the target type if it's a subtype of the current type
-                if (isTypeSubtypeOf(narrowedTypeCandidate, originalType)) {
-                    return narrowedTypeCandidate;
-                }
-                // If the current type is a union type, remove all constituents that aren't subtypes of the target.
+                // If the current type is a union type, remove all constituents that aren't assignable to target. If that produces
+                // 0 candidates, fall back to the assignability check
                 if (originalType.flags & TypeFlags.Union) {
-                    return getUnionType(filter((<UnionType>originalType).types, t => isTypeSubtypeOf(t, narrowedTypeCandidate)));
+                    let assignableConsituents = filter((<UnionType>originalType).types, t => isTypeAssignableTo(t, narrowedTypeCandidate));
+                    if (assignableConsituents.length) {

```

Comment:
I think this length check makes sense. It is never helpful to give the user empty object type just because we ran out of constituents.